### PR TITLE
VolumeControl: add controls for Effects and Music

### PR DESCRIPTION
**Pull request by UselessToucan**
It is now possible to adjust volume for effects and music.
osu_volume

### CODE REVIEW

File: osu.Game/VolumeControl.cs [hunk 1]
```diff
@@ -85,23 +72,37 @@ protected override bool OnWheelUp(InputState state)
         {
             appear();
 
-            VolumeGlobal.Value += 0.05f;
+            if (VolumeMeterSample.Contains(state.Mouse.Position))
```
Comment (peppy) [inline osu.Game/VolumeControl.cs]: rather than using contains logic here, i'd rather you moved it inside VolumeControl (and pass through the bindable).

**Review by ddevault — CHANGES_REQUESTED** [review]
If you use the scroll wheel while hovering over no meter in particular, it should adjust all the meters at once.

**Comment by peppy** [comment]
i think you mean master only >.>

**Comment by SirCmpwn** [comment]
Heh, sure.

**Comment by UselessToucan** [comment]
I didn't manage to do it in a different way, don't know is it acceptable or not. I would be glad if you explain me how to do it better.

**Comment by SirCmpwn** [comment]
You've gotta solve merge conflicts and check the build status.

File: osu.Game/VolumeMeter.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        {
+            base.Load();
+            updateFill();
+        }
+
+        protected override bool OnWheelUp(InputState state)
+        {
+            volume.Value += 0.05f;
+            updateFill();
+            return base.OnWheelUp(state);
```
Comment (peppy) [inline osu.Game/VolumeMeter.cs]: you want to return true here. return true when you handle the event (then it won't propagate further).

File: osu.Game/VolumeMeter.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        }
+
+        protected override bool OnWheelDown(InputState state)
+        {
+            volume.Value -= 0.05f;
+            updateFill();
+            return base.OnWheelDown(state);
```
Comment (peppy) [inline osu.Game/VolumeMeter.cs]: same here

File: osu.Game/VolumeControl.cs [hunk 4]
```diff
@@ -30,79 +25,45 @@ public override void Load(BaseGame game)
...
         }
 
         protected override bool OnWheelUp(InputState state)
         {
             appear();
-
-            VolumeGlobal.Value += 0.05f;
-            updateFill();
-
+            if (volumeMetersContainer.Children.All(vm => !vm.Contains(state.Mouse.Position)))
```
Comment (UselessToucan) [inline osu.Game/VolumeControl.cs]: Yes, it does not work as expected without these checks.

File: osu.Game/VolumeControl.cs [hunk 5]
```diff
@@ -25,84 +20,55 @@ public VolumeControl()
...
 
+        private void volumeChanged(object sender, System.EventArgs e)
+        {
+            appear();
+        }
+
         public override void Load(BaseGame game)
         {
             base.Load(game);
+            VolumeGlobal.ValueChanged += volumeChanged;
```
Comment (peppy) [inline osu.Game/VolumeControl.cs]: Please override dispose and unbind these.

File: osu.Game/VolumeMeter.cs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+                        meterFill = new Box
+                        {
+                            Colour = Color4.White,
+                            RelativeSizeAxes = Axes.Both,
+                            Origin = Anchor.BottomCentre,
+                            Anchor = Anchor.BottomCentre
+                        }
+                    }
+                },
+                new SpriteText {Text = meterName, Anchor = Anchor.BottomCentre,Origin = Anchor.BottomCentre,Position = new Vector2(0,-20)}
```
Comment (huoyaoyuan) [inline osu.Game/VolumeMeter.cs]: Consider to make this tidy

Comment (UselessToucan) [inline osu.Game/VolumeMeter.cs]: Should it be like this?
```
new SpriteText
{
    Text = meterName,
    Anchor = Anchor.BottomCentre,
    Origin = Anchor.BottomCentre,
    Position = new Vector2(0,-20)
}
```

Comment (SirCmpwn) [inline osu.Game/VolumeMeter.cs]: Yes

File: osu.Game/VolumeMeter.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                            Anchor = Anchor.BottomCentre
+                        }
+                    }
+                },
+                new SpriteText
+                {
+                    Text = meterName,
+                    Anchor = Anchor.BottomCentre,
+                    Origin = Anchor.BottomCentre,
+                    Position = new Vector2(0,-20)
```
Comment (peppy) [inline osu.Game/VolumeMeter.cs]: rather than use a position offset here, just set `Origin = Anchor.TopCentre`?

**Review by peppy — CHANGES_REQUESTED** [review]
just need the one fix, else looks good.

**Pull request merged**